## This is an individual assignment. Do not work in groups, and do not consult each other when doing this midterm.

# Task 1 - Getting tweets (3 points)

Pick either two Twitter usernames. These usernames should be "oppositional" in nature. I'm leaving the exact meaning of the term "oppositional" intentionally vague; the two usernames you pick should be in the same domain (politics, products, pop culture, etc.) but should also be strong contrasts or even rivals. For example:  @Apple vs. @SamsungMobile, @SenMajLdr vs. @SenSchumer, @CocaCola vs. @pepsi, @OfficialKanye vs. @taylorswift13. Basically, if the two handles/hashtags you pick can sensibly fit into the phrase "A vs. B," then they're oppositional.

In the code block below, use the twitter API to get **the text** of 500 tweets representing the two usernames you pick - specifically, the 500 most recent tweets of the users. Please pick users that tweet primarily in English. This "English" requirement is simply so the two instructors can go through the information. This means you'll have a total of 1000 tweets.

**Do NOT pick any of the examples I gave above above as your two oppositional usernames. Think up an original pair. Note that this task requires you to come up with two "oppositional" entities as well as to get those entities most recent tweets. Keep in mind that this is an individual assignment. Any two students who happen to pick the same two oppositional entities may be double checked for cheating, therefore, the more "original" your oppositional pair, the better.**

In [4]:
!pip install tweepy
import tweepy
import pandas as pd
from IPython.core.display import HTML

## get all the keys from the twitter console
api_key="****" ##enter your api key here
api_secret="****" ## enter your api secret here

## create the auth and O-auth objects
auth = tweepy.AppAuthHandler(api_key, api_secret)
api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

##initializing the variables
tweet_id_nike,tweet_text_nike=[],[]
tweet_id_adidas,tweet_text_adidas=[],[]

##getting all the tweets from the nike account
counter=0
print("Tweets from Nike account")
for page_no in range(1,26):
    user_tweets = api.user_timeline(id="Nike",lang="en",page=page_no)
    for tweet in reversed(user_tweets):
        tweet_id_nike.append(tweet.created_at)
        tweet_text_nike.append(tweet.text)
        if(counter>=500):
            break
        else:
            counter+=1

##print nike account tweets
df = pd.DataFrame({'created_at':tweet_id_nike, 'tweet':tweet_text_nike})
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(df)

##getting all the tweets from adidas account
counter=0
print("\nTweets from adidas account")
for page_no in range(1,26):
    user_tweets = api.user_timeline(id="adidas",lang="en",page=page_no)
    for tweet in reversed(user_tweets):
        tweet_id_adidas.append(tweet.created_at)
        tweet_text_adidas.append(tweet.text)
        if(counter>=500):
            break
        else:
            counter+=1

##print the adidas account tweets
df = pd.DataFrame({'created_at':tweet_id_adidas, 'tweet':tweet_text_adidas})
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(df)

Tweets from Nike account


created_at  \
0   2019-02-26 11:43:51   
1   2019-02-27 03:38:24   
2   2019-02-27 10:48:26   
3   2019-02-27 16:29:00   
4   2019-02-27 18:21:27   
5   2019-02-27 19:23:56   
6   2019-02-28 01:15:09   
7   2019-02-28 16:46:53   
8   2019-03-01 15:57:25   
9   2019-03-01 16:20:40   
10  2019-03-02 00:21:27   
11  2019-03-02 18:27:28   
12  2019-03-03 15:57:31   
13  2019-03-03 16:21:25   
14  2019-03-03 17:41:27   
15  2019-03-04 11:02:23   
16  2019-03-04 15:00:45   
17  2019-03-04 15:29:30   
18  2019-03-04 15:58:00   
19  2019-03-05 08:29:56   
20  2019-02-20 20:34:53   
21  2019-02-20 21:16:27   
22  2019-02-21 09:21:49   
23  2019-02-21 11:08:40   
24  2019-02-21 16:46:23   
25  2019-02-22 11:03:24   
26  2019-02-22 13:34:30   
27  2019-02-22 18:10:53   
28  2019-02-22 18:31:26   
29  2019-02-22 19:25:56   
30  2019-02-23 13:31:25   
31  2019-02-24 16:38:00   
32  2019-02-24 17:30:13   
33  2019-02-24 19:16:54   
34  2019-02-25 09:05:25   
35  2019-02-25 16:17:00   
36  2019-02-25 21:14:00   
37  2019-02-26 08:36:29   
38  2019-02-26 08:42:56   
39  2019-02-26 11:14:57   
40  2019-02-16 15:11:00   
41  2019-02-16 15:24:26   
42  2019-02-16 18:37:00   
43  2019-02-17 15:15:32   
44  2019-02-17 16:17:28   
45  2019-02-17 17:27:24   
46  2019-02-18 10:45:00   
47  2019-02-18 13:23:28   
48  2019-02-18 14:57:56   
49  2019-02-18 15:11:27   
50  2019-02-18 18:13:32   
51  2019-02-18 18:20:29   
52  2019-02-18 19:51:23   
53  2019-02-18 20:30:34   
54  2019-02-19 16:50:40   
55  2019-02-19 17:57:52   
56  2019-02-19 18:12:24   
57  2019-02-19 19:30:34   
58  2019-02-19 22:57:27   
59  2019-02-20 00:39:00   
60  2019-02-11 23:27:52   
61  2019-02-12 01:35:27   
62  2019-02-12 02:41:56   
63  2019-02-12 08:32:28   
64  2019-02-12 21:54:52   
65  2019-02-13 16:27:56   
66  2019-02-13 17:14:52   
67  2019-02-13 20:51:52   
68  2019-02-14 11:25:56   
69  2019-02-14 13:14:29   
70  2019-02-14 20:45:56   
71  2019-02-14 22:36:56   
72  2019-02-15 02:13:29   
73  2019-02-15 09:20:40   
74  2019-02-15 16:18:29   
75  2019-02-15 23:39:52   
76  2019-02-16 02:33:56   
77  2019-02-16 13:39:00   
78  2019-02-16 14:29:27   
79  2019-02-16 15:07:28   
80  2019-02-07 02:03:56   
81  2019-02-07 08:12:00   
82  2019-02-07 11:53:25   
83  2019-02-07 13:26:51   
84  2019-02-07 13:36:26   
85  2019-02-07 16:07:00   
86  2019-02-07 18:55:29   
87  2019-02-07 19:30:00   
88  2019-02-07 19:54:56   
89  2019-02-07 23:23:48   
90  2019-02-08 03:35:40   
91  2019-02-08 11:03:28   
92  2019-02-08 16:24:52   
93  2019-02-09 13:45:30   
94  2019-02-09 16:31:00   
95  2019-02-09 18:31:00   
96  2019-02-10 15:28:27   
97  2019-02-10 16:45:33   
98  2019-02-11 11:52:03   
99  2019-02-11 22:01:56   
100 2019-02-01 22:58:30   
101 2019-02-02 18:33:25   
102 2019-02-02 19:04:53   
103 2019-02-02 19:05:40   
104 2019-02-02 21:45:00   
105 2019-02-03 13:45:56   
106 2019-02-03 16:05:27   
107 2019-02-04 03:22:04   
108 2019-02-04 10:23:53   
109 2019-02-04 11:55:30   
110 2019-02-04 13:37:56   
111 2019-02-04 14:23:00   
112 2019-02-04 15:16:00   
113 2019-02-04 15:17:56   
114 2019-02-05 15:33:56   
115 2019-02-06 00:49:58   
116 2019-02-06 14:15:30   
117 2019-02-06 15:54:25   
118 2019-02-06 20:09:00   
119 2019-02-07 00:08:49   
120 2019-01-22 16:53:29   
121 2019-01-23 17:15:34   
122 2019-01-23 19:56:50   
123 2019-01-23 23:25:56   
124 2019-01-24 14:31:09   
125 2019-01-25 02:59:56   
126 2019-01-25 08:40:29   
127 2019-01-25 22:44:23   
128 2019-01-25 23:16:54   
129 2019-01-26 13:58:55   
130 2019-01-27 00:26:54   
131 2019-01-27 14:02:30   
132 2019-01-27 19:02:49   
133 2019-01-28 10:07:29   
134 2019-01-29 15:59:25   
135 2019-01-29 23:54:40   
136 2019-01-30 09:23:54   
137 2019-01-31 20:23:00   
138 2019-01-31 23:49:00   
139 2019-02-01 15:31:00   
140 2019-01-15 11:41:28   
141 2019-01-15 13:01:27   
142 2019-01-15 13:25:00   
143 2019-01-15 14:29:35   
144 2019-01-15 15:07:56   
145 2019-01-15 22:51:29   
146 2019-01-16 18:58:55   
147 2019-01-16 23


Tweets from adidas account


created_at  \
0   2019-01-07 21:46:14   
1   2019-01-07 22:05:01   
2   2019-01-08 21:08:01   
3   2019-01-08 21:14:42   
4   2019-01-08 21:29:57   
5   2019-01-08 21:34:21   
6   2019-01-08 21:36:42   
7   2019-01-08 21:37:56   
8   2019-01-08 21:40:11   
9   2019-01-14 19:11:23   
10  2019-01-14 19:12:03   
11  2019-01-25 12:53:33   
12  2019-01-26 11:14:15   
13  2019-01-26 11:25:43   
14  2019-02-02 23:58:20   
15  2019-02-04 03:22:52   
16  2019-02-11 05:57:07   
17  2019-02-11 16:13:30   
18  2019-02-25 15:20:00   
19  2019-02-27 08:30:27   
20  2018-12-11 18:04:54   
21  2018-12-11 18:44:09   
22  2018-12-15 00:29:37   
23  2018-12-19 07:02:53   
24  2019-01-07 20:47:46   
25  2019-01-07 20:49:49   
26  2019-01-07 20:53:14   
27  2019-01-07 20:54:53   
28  2019-01-07 21:00:10   
29  2019-01-07 21:03:10   
30  2019-01-07 21:06:23   
31  2019-01-07 21:10:48   
32  2019-01-07 21:12:34   
33  2019-01-07 21:15:05   
34  2019-01-07 21:19:00   
35  2019-01-07 21:29:13   
36  2019-01-07 21:33:56   
37  2019-01-07 21:38:22   
38  2019-01-07 21:40:39   
39  2019-01-07 21:42:21   
40  2018-11-06 18:24:04   
41  2018-11-06 18:26:04   
42  2018-11-07 15:44:08   
43  2018-11-08 23:07:35   
44  2018-11-09 22:57:15   
45  2018-11-10 20:28:29   
46  2018-11-17 16:24:27   
47  2018-11-17 16:35:14   
48  2018-11-17 20:31:00   
49  2018-12-07 17:00:44   
50  2018-12-07 18:04:29   
51  2018-12-07 18:14:42   
52  2018-12-09 04:59:19   
53  2018-12-10 02:24:22   
54  2018-12-10 19:06:15   
55  2018-12-10 19:50:33   
56  2018-12-10 20:49:38   
57  2018-12-10 20:51:34   
58  2018-12-11 17:09:42   
59  2018-12-11 17:35:36   
60  2018-09-21 06:08:50   
61  2018-10-08 12:19:41   
62  2018-10-23 15:21:37   
63  2018-10-25 20:05:15   
64  2018-10-26 02:51:13   
65  2018-10-29 03:27:40   
66  2018-10-29 03:31:55   
67  2018-10-29 03:38:32   
68  2018-10-29 21:01:50   
69  2018-11-04 16:47:16   
70  2018-11-05 12:42:03   
71  2018-11-06 18:06:04   
72  2018-11-06 18:08:04   
73  2018-11-06 18:10:04   
74  2018-11-06 18:12:04   
75  2018-11-06 18:14:04   
76  2018-11-06 18:16:04   
77  2018-11-06 18:18:04   
78  2018-11-06 18:20:04   
79  2018-11-06 18:22:04   
80  2018-08-28 12:58:05   
81  2018-08-28 13:00:05   
82  2018-08-28 13:02:04   
83  2018-08-28 13:04:04   
84  2018-08-28 13:06:05   
85  2018-08-28 13:08:07   
86  2018-08-28 13:10:08   
87  2018-08-28 13:12:04   
88  2018-08-28 13:14:05   
89  2018-08-28 13:26:04   
90  2018-08-28 13:28:04   
91  2018-08-28 13:30:04   
92  2018-08-28 13:32:04   
93  2018-08-28 13:34:04   
94  2018-09-06 09:05:01   
95  2018-09-08 21:54:37   
96  2018-09-16 11:19:00   
97  2018-09-18 14:48:26   
98  2018-09-18 16:01:52   
99  2018-09-19 06:34:00   
100 2018-08-24 21:38:04   
101 2018-08-24 21:40:04   
102 2018-08-24 21:42:04   
103 2018-08-24 21:44:04   
104 2018-08-24 21:46:04   
105 2018-08-24 21:48:04   
106 2018-08-24 21:50:04   
107 2018-08-24 21:52:04   
108 2018-08-24 21:54:04   
109 2018-08-24 21:56:04   
110 2018-08-27 15:20:05   
111 2018-08-28 10:39:00   
112 2018-08-28 12:40:04   
113 2018-08-28 12:44:04   
114 2018-08-28 12:46:06   
115 2018-08-28 12:48:04   
116 2018-08-28 12:50:04   
117 2018-08-28 12:52:05   
118 2018-08-28 12:54:04   
119 2018-08-28 12:56:04   
120 2018-08-24 20:58:06   
121 2018-08-24 21:00:05   
122 2018-08-24 21:02:04   
123 2018-08-24 21:04:04   
124 2018-08-24 21:06:04   
125 2018-08-24 21:08:04   
126 2018-08-24 21:10:04   
127 2018-08-24 21:12:04   
128 2018-08-24 21:14:04   
129 2018-08-24 21:16:05   
130 2018-08-24 21:18:04   
131 2018-08-24 21:20:04   
132 2018-08-24 21:22:04   
133 2018-08-24 21:24:04   
134 2018-08-24 21:26:05   
135 2018-08-24 21:28:04   
136 2018-08-24 21:30:04   
137 2018-08-24 21:32:05   
138 2018-08-24 21:34:04   
139 2018-08-24 21:36:04   
140 2018-08-23 21:14:04   
141 2018-08-23 21:16:04   
142 2018-08-23 21:18:04   
143 2018-08-24 20:24:04   
144 2018-08-24 20:26:05   
145 2018-08-24 20:28:04   
146 2018-08-24 20:30:04   
147 2018-08-24 20

## Task 1 Rubric

- code to obtain Tweets runs without errors (1 point)
- code correctly written to obtain the latest 500 tweets from a pair of usernames (1 point)
- student REMOVES their API key and secret before submitting the midterm!!! (1 point)

# Task 2 - Saving your data to an external file (3 points)

In the code block below, consolidate your tweets into a single variable. This variable should have two "columns," one for the text of the tweets, the other a binary indicator of the source (e.g. 0 for the first source, 1 for the second). You can use a list of tuples or A Pandas dataframe, and save as either a pickle (.pkl) or a comma-seperated values file (.csv).

In [22]:
## 0 for nike and 1 for adidas
print("Exporting data to files...........")
combined_tweets,labels=[],[]
for i in range(500):
    combined_tweets.append(tweet_text_nike[i])
    labels.append("0")
for i in range(500):
    combined_tweets.append(tweet_text_adidas[i])
    labels.append("1")

##using a pandas dataframe
df=pd.DataFrame({'Source':labels, 'Tweets':combined_tweets}) ## one single variable of pandas dataframe with two columns tweets and source
df.to_csv('csv_file.csv',columns=["Tweets","Source"],index=False) ## saving it to csv file
print("Success! Exporting data to files")

Exporting data to files...........
Success! Exporting data to files


## Task 2 Rubric

- code to save data to external file runs without errors (1 point)
- saved data formatted correctly into "2 columns" (1 point)
- external file submitted with midterm (1 point)

# Task 3 - Preparing data for an sklearn binary classifier (3 points)

In the code block below, create two variables, X, and y. The y variable should be simple - it is simply the "second column" of the data you made in task 2, a binary indicator of source, with 0 representing one source and 1 representing the other.

For the X variable, choose either `TfidifVectorizer` or `CountVectorizer` from `sklearn.feature_representation.text` to turn the raw text (column 1 from task 2) into a "bag-of-words" representation. When instantiating your vectorizer, set the argument `lowercase=True`, to ensure that all words are lowercased, and `stopwords="english"`, to remove English stop words. 

Additional, when instantiating the vectorizer, pass the `max_df=???` and `min_df=???` arguments. These arguments can either take a *float between 0.0 and 1.0 or an integer*. The df stands for "document frequency." These arguments tell the vectorizer to remove words that occur *over* (max_df) and *under* (min_df) a certain amount of documents. This will remove frequent words - which show up all the time and therefore are not informative - and infrequent words, which are so rare as to just be noise. If you pass these arguments a float, that float represents the proportion of documents (e.g. `max_df=0.9` means, remove all words that show up in more than 90% of the tweets) and if you pass these arguments an integer, that integer represents the number of documents (e.g. `min_df=5`, remove all words that show up in less than 5 document). 

By the end of this task, you should have a variable X, of dimensionality $n \times d$ where $n = 1000$ and $d$ is the number of words left after the vectorizer considers df, and you should have variable y, which is a vector of length 1000, with 1s or 0s representing tweet source.

In [23]:
from sklearn.feature_extraction.text import TfidfVectorizer
df2=pd.read_csv("csv_file.csv")
df2.columns.values.tolist()
tweet_col=df2['Tweets']
y=df2['Source']

vectorizer = TfidfVectorizer(lowercase=True, stop_words="english",max_df=100, min_df=5)
X = vectorizer.fit_transform(tweet_col)
print("Shape of X variable: ",X.shape)
print("Shape of Y variable: ",y.shape," length of Y variable: ",len(y))

Shape of X variable:  (1000, 237)
Shape of Y variable:  (1000,)  length of Y variable:  1000


## Task 3 Rubric

- Code runs without errors (1 point)
- Written code correctly achieves objective of creating X, y variables for classifier (1 point)
- All required arguments to vectorizer included (1 point)


# Task 4 - Training a Logistic Regression classifier (3 points)

Instantiate an sklearn Logistic Regression binary classifier ([sklearn documentation here](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html)). 

Then, use `cross_val_score` from `sklearn.model_selection` ([sklearn documentation here](http://scikit-learn.org/stable/modules/cross_validation.html#computing-cross-validated-metrics))to perform 5-fold cross validation. The inputs to `cross_val_score` will be your instantiated Logistic Regression classifier, X, y, and a named argument `cv=5` to indicate the number of folds. The output will be a list of 5 numbers - the accuracy from each fold.

Print the average of those 5 numbers. This will be the mean 5-fold cross validation accuracy of your classifier.

In [25]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

clf = LogisticRegression(random_state=0, solver='lbfgs',multi_class='multinomial').fit(X, y)
scores = cross_val_score(clf,X,y,cv=5)
print("The mean cross validation accuracy of the classifier is (in %): ",scores.mean()*100)

The mean cross validation accuracy of the classifier is (in %):  81.89999999999999


## Task 4 Rubric

- Code runs without errors (1 point)
- Code successfully creates a Logistic Regression classifier and runs cross validation (1 point)
- Code prints mean cross validation accuracy (1 point)

# Task 5 - Discussion

Answer the following questions.

1 Since you pulled equal amounts of tweets from each source, the baseline accuracy is 50%. This is the accuracy we would expect from a classifier that guessed 0 or 1 randomly, or a classifier that simply guessed all 0s or all 1s. Your classifier either did well or did poorly. In either case, think about the *actual content* of the sources you picked and in the text block below, informally share your thoughts on why your classifier did poorly/well. (1 point)

The classifier works well on the classifying the tweets from the two accounts this was mainly because of:
1. The classifier had enough data to work on ie.. large enough data corpus which included of 500 tweets for each account, accounting to a total of 1000 tweets to work on. As usually classifiers need a lot of data to classify correctly.
2. The actual content of the tweets from both the accounts actually matter too, while collecting tweets from Nike and Adidas I noticed that nike twitter handle usually provided solutions or answers to their customers queries. Whereas the twitter handle of Adidas never actually provided any solution but instead gave replies like "we hear you", "we will get back to you" etc.
3. Not only the content for the nike tweets was more but also they had more of vocabulary and a range of action words, verbs as compared to the Adidas twitter handle.
This might just help the the classifier to classify the tweets correctly simply using the style of vocabulary in the tweets of both the accounts.

2 What could you have done differently when preprocessing your data (task 3) to try and improve your classifier's accuracy? (1 point)

Possible solutions to pre-process data in Task3 instead of simply using max_df and min_df to create the dictionary of words:
- remove numerical values from the tweet as they do not signify anything and adds up to the words if a certain number is used repeatedly
- try to remove typos in the tweet as while the bag of words had a lot of typos in it, this could have been done manually or readily available dictionary and tools online to remove slangs and typos in the tweet
- removing special characters such as "@" and emoticons as we are not analyzing any sentiments these can probably be removed from the tweet because a tweet can contain :) from both the accounts this confuses the classifier as to which account does it actually belong as it is not a significant feature to distinguish the 2 accounts.

3 What parameters could you have adjusted in the Logistic Regression classifier in Task 4 to "tune" it and get better performance? What other binary classifiers could you have used, and what "tune-able" parameters do those classifiers have? (1 point)

Parameters that could have been tuned:
- adding a penalty='l2'
- C=0.3 this is used for regularizing the data, smaller the C stronger is the regularization this would help us regularizing the data in the dataset
- using a different solver such as ‘newton-cg’, ‘liblinear’, ‘sag’, ‘saga’ instead of 'lbfgs'
- the 'cv' parameter in the cross_val_score, just using 5 folds isnt usually the best way to determine the accuracy machine learning classifier use atleast 200 samples or folds to determine the accuracy of a classifier

[1] Other binary classifiers that could have been used are:
- Decision trees
- Random forests
- Bayesian networks
- Support vector machines
- Neural networks
- Probit model
reference: [1] https://en.wikipedia.org/wiki/Binary_classification

# Bonus (3 points)

This is an opportunity to gain an additional 3 points above the 15 allocated for the midterm. This will require you to do something that wasn't covered directly in the lectures, but can somewhat easily be learned by going through the sklearn documentation.

Note 1: When a count vectorizer or TF-IDF vectorizer is instantiated and used to transform your raw text data, it builds a dictionary that indicates which word is assigned to which index. Remember that it produces an $n \times d$ matrix, where $n$ is the number of samples and $d$ is the number of words. If you want to know the index of a word (that is, which column in $d$ corresponds to that word), you can consult this dictionary. Suppose you named your vectorizer `vec`. To access this dictionary, use `vec.vocabulary_`. If you want to know the index of the word `banana`, access `vec.vocabulary_['banana']`. 


Note 2: When you instantiate and train a logistic regression, it saves a set of *coefficients* indicating the "weight" of that word in terms of predicting the outcome variable. Suppose you named your classifier `lr`. You can access these coefficients at `lr.coef_[0]`, which is a one-dimensional array. This means that `lr.coef_[0][0]` is the weight of the 0th feature, `lr.coef_[0][1]` is the weight of the 1st feature, and so on. You can also access the intercept at `lr.intercept_`

Therefore, you can *iterate* through `vec.vocabulary_.items()`, and for each word (key) get its index (value) and then find the coefficient weight of that word in the model `lr.coef_[0][index]`


In the code block (or blocks, if you want to make more than one to organize your code better) below:

1. Instantiate a *new* instance of a Logistic Regression classifier, `fit` that classifier on X and y. (1 point)
2. Use the notes above to make a list of tuples, where the first value in each tuple is a *word in the vocabulary* and the second value is the *coefficient weight assigned to that word in the trained Logistic Regression classifier*. Sort that list of tuples by the second value (the weight) ([Here's how you can do that](https://stackoverflow.com/questions/10695139/sort-a-list-of-tuples-by-2nd-item-integer-value)). (1 point)
3. Print the 10 words with the highest weights and the 10 words with the lowest weights. In a few sentences discuss whether these words help you understand why the model performed well/poorly. (1 point).

In [26]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn.feature_extraction.text import TfidfVectorizer
## read the data from the file and create the list
df2=pd.read_csv("csv_file.csv")
df2.columns.values.tolist()
tweet_col=df2['Tweets']
y=df2['Source']

##creating the vectorizer
vec = TfidfVectorizer(lowercase=True, max_df=100, min_df=5)
X = vec.fit_transform(tweet_col)

##creating the logistic regression
lr = LogisticRegression(random_state=0, solver='lbfgs',multi_class='multinomial').fit(X, y)

##creating the list with tuples
counter=0
my_list=[]
for x in vec.vocabulary_:
    tuple=(x, lr.coef_[0][counter])
    my_list.append(tuple)
    counter+=1

##sorting the list with the weight as given in the reference
my_list=sorted(my_list, key=lambda x: x[1])

##printing the top 10 words in a pandas table
print("Top 10 words with the highest weight are: ")
temp_list1, temp_list2=[], []
for i in my_list[-10:]:
    temp_list1.append(i[0])
    temp_list2.append(i[1])

##reversing the list so as to get the output in descending order 
temp_list1.reverse()
temp_list2.reverse()
df = pd.DataFrame({'Word':temp_list1, 'weight':temp_list2})
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(df)
    
##printing the top 10 words in a pandas table
print("Top 10 words with the lowest weight are: ")
temp_list1, temp_list2=[], []
for i in my_list[:10]:
    temp_list1.append(i[0])
    temp_list2.append(i[1])
    
df = pd.DataFrame({'Word':temp_list1, 'weight':temp_list2})
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(df)

Top 10 words with the highest weight are: 


Word    weight
0  covered     2.203546
1  news        1.577548
2  it          1.385370
3  air         1.241319
4  nice        1.180252
5  when        1.129970
6  don         1.096986
7  locked      1.086379
8  considered  1.086355
9  now         1.082263

Top 10 words with the lowest weight are: 


Word    weight
0  longer   -1.660505
1  day      -1.134612
2  isn      -1.024465
3  podrías  -1.018330
4  retailer -1.007191
5  pass     -0.983505
6  here     -0.969699
7  most     -0.963044
8  on       -0.939695
9  women    -0.850372

Part 3: In a few sentences discuss whether these words help you understand why the model performed well/poorly (1 point)

The model was able to perform well because the weighrs to the words have not been near the classification boundary as we can notice that most of the words are very distinct to a set of accout thus the weights assigned to them have been away from the decision boundary(which is 0 in this case). Greater the value of weights easier is to classify the data from the classifier. 
Also these words carried the highest weightage in an account if a word with weight (2.2) is present in the tweet then that tweet has a higher probability of getting classified on the right side of the 0 boundary rather than the word having a weight of (0.05) which can be misclassified to the left of the decision boundary.